utils/bitfield/src/iter/combine.rs

```diff
@@ -0,0 +1,492 @@
+where
+    A: RangeIterator,
+    B: RangeIterator,
+    C: Combinator,
+{
+    pub fn new(a: A, b: B) -> Self {
+        Self(Merge::new(_Combine::new(a, b)))
+    }
+}
+
+impl<A, B, C> Iterator for Combine<A, B, C>
+where
+    A: RangeIterator,
+    B: RangeIterator,
+    C: Combinator,
+{
+    type Item = Range<usize>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next()
+    }
+}
+
+impl<A, B, C> RangeIterator for Combine<A, B, C>
+where
+    A: RangeIterator,
+    B: RangeIterator,
+    C: Combinator,
+{
+}
+
+/// Combines two range iterators according to the given combinator, but does not
+/// merge the output ranges together. Since the ranges can overlap, this does not
+/// satisfy the `RangeIterator` requirements.
+struct _Combine<A, B, C>
+where
+    A: RangeIterator,
+    B: RangeIterator,
+{
+    lhs: Lookahead<A>,
+    rhs: Lookahead<B>,
+    combinator: C,
+}
+
+impl<A, B, C> _Combine<A, B, C>
+where
+    A: RangeIterator,
+    B: RangeIterator,
+    C: Combinator,
+{
+    fn new(lhs: A, rhs: B) -> Self {
+        Self {
+            lhs: Lookahead::new(lhs),
+            rhs: Lookahead::new(rhs),
+            combinator: Default::default(),
+        }
+    }
+
+    /// Computes the next range by inspecting the next range of each of the input
+    /// range iterators and passing them to the combinator. Also advances the range
+    /// iterator which corresponding range has the lowest upper bound.
+    fn next_range(&mut self) -> Option<Range<usize>> {
+        let (range, advance_lhs) = match (self.lhs.peek(), self.rhs.peek()) {
+            (Some(lhs), Some(rhs)) => {
+                // if both iterators are non-empty, we advance the one whichever's
+                // corresponding range ends first
+                if lhs.end <= rhs.end {
+                    (Some(self.combinator.advance_lhs(lhs.clone(), rhs)), true)
+                } else {
+                    (Some(self.combinator.advance_rhs(lhs, rhs.clone())), false)
+                }
+            }
+            (Some(lhs), None) => (self.combinator.advance_lhs_tail(lhs.clone()), true),
+            (None, Some(rhs)) => (self.combinator.advance_rhs_tail(rhs.clone()), false),
+            (None, None) => return None,
+        };
+
+        if advance_lhs {
+            self.lhs.next();
+        } else {
+            self.rhs.next();
+        }
+
+        range
+    }
+}
+
+impl<A, B, C> Iterator for _Combine<A, B, C>
+where
+    A: RangeIterator,
+    B: RangeIterator,
+    C: Combinator,
+{
+    type Item = Range<usize>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // we repeatedly compute the next range until we find one that is non-empty
+        // TODO: use `!range.is_empty()` once it stabilizes in Rust 1.47
+        iter::from_fn(|| self.next_range()).find(|range| range.start < range.end)
+    }
+}
+
+impl<A, B, C> RangeIterator for _Combine<A, B, C>
+where
+    A: RangeIterator,
+    B: RangeIterator,
+    C: Combinator,
+{
+}
+
```

Comment:
```suggestion
```
You explicitly say this is not satisfied. (I'm still a bit confused about why _Combine is even necessary) You are wrapping it as `Combine(Merge(_Combine))` but does this just remove some code duplication or something?

Reply:
Yes, great catch, that is indeed not satisfied. `Combine` does implement it, `_Combine` doesn't.

The `_Combine` layering is less than ideal, if trait methods could return `impl Trait` then there'd be no need for a type that wraps `Merge<_Combine>`. I found that integrating the `Merge` code into what is now `_Combine` made it a lot messier, and `Merge` will indeed be able to be used elsewhere in the future (in unioning multiple bitfields, which is currently implemented very naively).